### PR TITLE
Executor - Save benchmark results to file

### PR DESCRIPTION
**Pull request by abuccts**
__Major Revision__

* Save benchmark results to json file.

### CODE REVIEW

File: superbench/executor/executor.py [hunk 1]
```diff
@@ -106,17 +110,47 @@ def __exec_benchmark(self, context, log_suffix):
...
+        Args:
+            benchmark_name (str): Benchmark name.
+            create (bool): Create the directory or not.
+
+        Return:
+            str: Benchmark output directory.
+        """
+        benchmark_output_dir = Path(self._output_dir, 'benchmarks', benchmark_name)
+        if create:
+            benchmark_output_dir.mkdir(mode=0o755, parents=True, exist_ok=True)
```
Comment (cp5555) [inline superbench/executor/executor.py:133]: Do we need to empty the dir?

File: superbench/executor/executor.py [hunk 2]
```diff
@@ -106,17 +110,47 @@ def __exec_benchmark(self, context, log_suffix):
...
+        return str(benchmark_output_dir)
+
+    def __write_benchmark_results(self, benchmark_name, benchmark_results):
+        """Write benchmark results.
+
+        Args:
+            benchmark_name (str): Benchmark name.
+            benchmark_results (dict): Benchmark results.
+        """
+        with Path(self.__get_benchmark_dir(benchmark_name), 'results.json').open(mode='w') as f:
```
Comment (cp5555) [inline superbench/executor/executor.py:143]: Is it a separate json file?

File: superbench/executor/executor.py (hunk 1 above)
Comment (abuccts) [reply superbench/executor/executor.py:133]: nope? if the same output dir is used, later output will rewrite previous one

File: superbench/executor/executor.py (hunk 2 above)
Comment (abuccts) [reply superbench/executor/executor.py:143]: every benchmark will have a results.json in its directory
we may need a merged json after runner collected all json on all nodes

Comment (cp5555) [reply superbench/executor/executor.py:143]: I think one json should be good.

File: superbench/executor/executor.py [hunk 3]
```diff
@@ -106,17 +110,47 @@ def __exec_benchmark(self, context, log_suffix):
...
+        """Get output directory for benchmark.
+
+        Args:
+            benchmark_name (str): Benchmark name.
+            create (bool): Create the directory or not.
+
+        Return:
+            str: Benchmark output directory.
+        """
+        benchmark_output_dir = Path(self._output_dir, 'benchmarks', benchmark_name)
```
Comment (guoshzhao) [inline superbench/executor/executor.py:131]: Each run will have different `self._output_dir`, right?

File: superbench/executor/executor.py [hunk 4]
```diff
@@ -106,17 +110,47 @@ def __exec_benchmark(self, context, log_suffix):
...
+                return benchmark.result
             else:
                 logger.error('Executor failed in %s, invalid context.', log_suffix)
-        except Exception:
+        except Exception as e:
+            logger.error(e)
             logger.error('Executor failed in %s.', log_suffix)
+        return None
+
+    def __get_benchmark_dir(self, benchmark_name, create=False):
```
Comment (guoshzhao) [inline superbench/executor/executor.py:121]: Why need one option for `create`?

File: superbench/executor/executor.py [hunk 5]
```diff
@@ -106,17 +110,47 @@ def __exec_benchmark(self, context, log_suffix):
                     logger.info('Executor succeeded in %s.', log_suffix)
                 else:
                     logger.error('Executor failed in %s.', log_suffix)
+                return benchmark.result
```
Comment (guoshzhao) [inline superbench/executor/executor.py:113]: benchmark.result is only the summary data, we'd better return all info of benchmark such as raw_data and meta_data (start/end time, return code, benchmark name etc.). How about:
`return json.loads(benchmark.serialized_result)`

File: superbench/executor/executor.py (hunk 2 above)
Comment (cp5555) [reply superbench/executor/executor.py:143]: Besides, based on our previous bad experience, if we do the merge offline. We don't need to do this in this version. We should do it in the runner.

File: superbench/executor/executor.py (hunk 1 above)
Comment (abuccts) [reply superbench/executor/executor.py]: added directory rotation and keep it empty

File: superbench/executor/executor.py (hunk 2 above)
Comment (abuccts) [reply superbench/executor/executor.py]: will add in runner in another pr

File: superbench/executor/executor.py (hunk 3 above)
Comment (abuccts) [reply superbench/executor/executor.py:128]: yes, it's create here
https://github.com/microsoft/superbenchmark/blob/f1327bb788aeda7d755ee102e751a8b0edb038fa/superbench/common/utils/file_handler.py#L20

File: superbench/executor/executor.py (hunk 4 above)
Comment (abuccts) [reply superbench/executor/executor.py]: removed

File: superbench/executor/executor.py (hunk 5 above)
Comment (abuccts) [reply superbench/executor/executor.py]: updated

**Review by cp5555 — APPROVED** [review]

File: tests/executor/test_executor.py [hunk 6]
```diff
@@ -74,6 +75,43 @@ def test_get_arguments(self):
...
+        self.executor._SuperBenchExecutor__create_benchmark_dir('foobar')
+        foobar_results = {
+            'sum': 1,
+            'avg': 1.1,
+        }
+        self.executor._SuperBenchExecutor__write_benchmark_results('foobar', foobar_results)
+        self.assertTrue(foobar_results_path.is_file())
+        with foobar_results_path.open(mode='r') as f:
+            self.assertDictEqual(json.load(f), foobar_results)
+
```
Comment (cp5555) [inline tests/executor/test_executor.py:114]: Please improve the code coverage.

**Review by guoshzhao — APPROVED** [review]

**Pull request merged**